fnmpapi.lib should (optionally?) be a DLL

Statically linking to the FNMPAPI triggers some linker warnings about default libraries. It would probably be best to package the API wrappers into a DLL or perhaps even just a header file.

```
         LINK : warning LNK4098: defaultlib 'LIBCMT' conflicts with use of other libs; use /NODEFAULTLIB:library [D:\test\functional\functionaltests.vcxproj]
```

--- a/inc/fnmpapi.h
+++ b/inc/fnmpapi.h
@@ -9,33 +9,42 @@ EXTERN_C_START
 
 #include <fniotypes.h>
 
+#ifndef FNMPAPI
+#define FNMPAPI __declspec(dllimport)
+#endif
+
 #define FNMP_DEFAULT_RSS_QUEUES 4
 #define FNMP_MAX_RSS_INDIR_COUNT 128
 
+FNMPAPI
 HRESULT
 FnMpOpenShared(
     _In_ UINT32 IfIndex,
     _Out_ HANDLE *Handle
     );
 
+FNMPAPI
 HRESULT
 FnMpOpenExclusive(
     _In_ UINT32 IfIndex,
     _Out_ HANDLE *Handle
     );
 
+FNMPAPI
 HRESULT
 FnMpRxEnqueue(
     _In_ HANDLE Handle,
     _In_ DATA_FRAME *Frame
     );
 
+FNMPAPI
 HRESULT
 FnMpRxFlush(
     _In_ HANDLE Handle,
     _In_opt_ DATA_FLUSH_OPTIONS *Options
     );
 
+FNMPAPI
 HRESULT
 FnMpTxFilter(
     _In_ HANDLE Handle,
@@ -44,6 +53,7 @@ FnMpTxFilter(
     _In_ UINT32 Length
     );
 
+FNMPAPI
 HRESULT
 FnMpTxGetFrame(
     _In_ HANDLE Handle,
@@ -53,17 +63,20 @@ FnMpTxGetFrame(
     _Out_opt_ DATA_FRAME *Frame
     );
 
+FNMPAPI
 HRESULT
 FnMpTxDequeueFrame(
     _In_ HANDLE Handle,
     _In_ UINT32 FrameIndex
     );
 
+FNMPAPI
 HRESULT
 FnMpTxFlush(
     _In_ HANDLE Handle
     );
 
+FNMPAPI
 HRESULT
 FnMpGetLastMiniportPauseTimestamp(
     _In_ HANDLE Handle,
@@ -74,12 +87,14 @@ FnMpGetLastMiniportPauseTimestamp(
 #define FNMP_MAX_MTU (16 * 1024 * 1024)
 #define FNMP_DEFAULT_MTU FNMP_MAX_MTU
 
+FNMPAPI
 HRESULT
 FnMpSetMtu(
     _In_ HANDLE Handle,
     _In_ UINT32 Mtu
     );
 
+FNMPAPI
 HRESULT
 FnMpOidFilter(
     _In_ HANDLE Handle,
@@ -87,6 +102,7 @@ FnMpOidFilter(
     _In_ UINT32 KeyCount
     );
 
+FNMPAPI
 HRESULT
 FnMpOidGetRequest(
     _In_ HANDLE Handle,
@@ -95,6 +111,7 @@ FnMpOidGetRequest(
     _Out_opt_ VOID *InformationBuffer
     );
 
+FNMPAPI
 HRESULT
 FnMpOidCompleteRequest(
     _In_ HANDLE Handle,

--- a/src/lib/fnmpapi.c
+++ b/src/lib/fnmpapi.c
@@ -5,6 +5,7 @@
 
 #include "precomp.h"
 
+FNMPAPI
 HRESULT
 FnMpOpenShared(
     _In_ UINT32 IfIndex,
@@ -20,6 +21,7 @@ FnMpOpenShared(
     return FnMpOpen(FILE_CREATE, EaBuffer, sizeof(EaBuffer), Handle);
 }
 
+FNMPAPI
 HRESULT
 FnMpOpenExclusive(
     _In_ UINT32 IfIndex,
@@ -35,6 +37,7 @@ FnMpOpenExclusive(
     return FnMpOpen(FILE_CREATE, EaBuffer, sizeof(EaBuffer), Handle);
 }
 
+FNMPAPI
 HRESULT
 FnMpRxEnqueue(
     _In_ HANDLE Handle,
@@ -53,6 +56,7 @@ FnMpRxEnqueue(
     return FnMpIoctl(Handle, IOCTL_RX_ENQUEUE, &In, sizeof(In), NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpRxFlush(
     _In_ HANDLE Handle,
@@ -72,6 +76,7 @@ FnMpRxFlush(
     return FnMpIoctl(Handle, IOCTL_RX_FLUSH, &In, sizeof(In), NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpTxFilter(
     _In_ HANDLE Handle,
@@ -101,6 +106,7 @@ FnMpTxFilter(
     return FnMpIoctl(Handle, IOCTL_TX_FILTER, &In, sizeof(In), NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpTxGetFrame(
     _In_ HANDLE Handle,
@@ -142,6 +148,7 @@ FnMpTxGetFrame(
     return Result;
 }
 
+FNMPAPI
 HRESULT
 FnMpTxDequeueFrame(
     _In_ HANDLE Handle,
@@ -160,6 +167,7 @@ FnMpTxDequeueFrame(
     return FnMpIoctl(Handle, IOCTL_TX_DEQUEUE_FRAME, &In, sizeof(In), NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpTxFlush(
     _In_ HANDLE Handle
@@ -172,6 +180,7 @@ FnMpTxFlush(
     return FnMpIoctl(Handle, IOCTL_TX_FLUSH, NULL, 0, NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpGetLastMiniportPauseTimestamp(
     _In_ HANDLE Handle,
@@ -188,6 +197,7 @@ FnMpGetLastMiniportPauseTimestamp(
             sizeof(*Timestamp), NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpSetMtu(
     _In_ HANDLE Handle,
@@ -206,6 +216,7 @@ FnMpSetMtu(
     return FnMpIoctl(Handle, IOCTL_MINIPORT_SET_MTU, &In, sizeof(In), NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpOidFilter(
     _In_ HANDLE Handle,
@@ -224,6 +235,7 @@ FnMpOidFilter(
             Handle, IOCTL_OID_FILTER, (VOID *)Keys, sizeof(Keys[0]) * KeyCount, NULL, 0, NULL, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpOidGetRequest(
     _In_ HANDLE Handle,
@@ -247,6 +259,7 @@ FnMpOidGetRequest(
             *InformationBufferLength, InformationBufferLength, NULL);
 }
 
+FNMPAPI
 HRESULT
 FnMpOidCompleteRequest(
     _In_ HANDLE Handle,

--- a/src/lib/fnmpapi.vcxproj
+++ b/src/lib/fnmpapi.vcxproj
@@ -3,7 +3,9 @@
   <PropertyGroup>
     <ProjectGuid>{15de7d30-6a0b-47c7-8ddc-1cfc658e8a1e}</ProjectGuid>
     <TargetName>fnmpapi</TargetName>
-    <UndockedType>lib</UndockedType>
+    <!-- Temporary hack to disable linking stage in official OneBranch builds, because it's broken -->
+    <UndockedType Condition="'$(UndockedOfficial)' != 'true'">dll</UndockedType>
+    <UndockedType Condition="'$(UndockedOfficial)' == 'true'">lib</UndockedType>
     <UndockedDir>$(SolutionDir)submodules\undocked\</UndockedDir>
     <UndockedOut>$(SolutionDir)artifacts\</UndockedOut>
     <UndockedSourceLink>true</UndockedSourceLink>
@@ -28,6 +30,13 @@
         %(PreprocessorDefinitions)
       </PreprocessorDefinitions>
     </ClCompile>
+    <Link>
+      <SubSystem>Windows</SubSystem>
+      <AdditionalDependencies>
+        ntdll.lib;
+        %(AdditionalDependencies)
+      </AdditionalDependencies>
+    </Link>
   </ItemDefinitionGroup>
   <Import Project="$(UndockedDir)vs\windows.undocked.targets" />
 </Project>

--- a/src/lib/fnmpapi.vcxproj
+++ b/src/lib/fnmpapi.vcxproj
@@ -3,9 +3,7 @@
   <PropertyGroup>
     <ProjectGuid>{15de7d30-6a0b-47c7-8ddc-1cfc658e8a1e}</ProjectGuid>
     <TargetName>fnmpapi</TargetName>
-    <!-- Temporary hack to disable linking stage in official OneBranch builds, because it's broken -->
-    <UndockedType Condition="'$(UndockedOfficial)' != 'true'">dll</UndockedType>
-    <UndockedType Condition="'$(UndockedOfficial)' == 'true'">lib</UndockedType>
+    <UndockedType>dll</UndockedType>
     <UndockedDir>$(SolutionDir)submodules\undocked\</UndockedDir>
     <UndockedOut>$(SolutionDir)artifacts\</UndockedOut>
     <UndockedSourceLink>true</UndockedSourceLink>
@@ -31,7 +29,6 @@
       </PreprocessorDefinitions>
     </ClCompile>
     <Link>
-      <SubSystem>Windows</SubSystem>
       <AdditionalDependencies>
         ntdll.lib;
         %(AdditionalDependencies)

--- a/src/lib/precomp.h
+++ b/src/lib/precomp.h
@@ -5,6 +5,8 @@
 
 #pragma once
 
+#define FNMPAPI __declspec(dllexport)
+
 #include <windows.h>
 #include <winioctl.h>
 #include <winternl.h>

--- a/tools/create-devkit.ps1
+++ b/tools/create-devkit.ps1
@@ -34,6 +34,6 @@ copy -Recurse inc\* $DstPath\include
 
 New-Item -Path $DstPath\lib -ItemType Directory > $null
 copy "artifacts\bin\$($WinArch)$($WinConfig)\fnmpapi.lib" $DstPath\lib
-# TODO - PBD for .lib? copy "artifacts\bin\$($WinArch)$($WinConfig)\fnmpapi.pdb" $DstPath\lib
+copy "artifacts\bin\$($WinArch)$($WinConfig)\fnmpapi.pdb" $DstPath\lib
 
 Compress-Archive -DestinationPath "$DstPath\$Name.zip" -Path $DstPath\*

--- a/tools/create-runtimekit.ps1
+++ b/tools/create-runtimekit.ps1
@@ -31,9 +31,11 @@ New-Item -Path $DstPath -ItemType Directory > $null
 
 New-Item -Path $DstPath\bin -ItemType Directory > $null
 copy -Recurse "artifacts\bin\$($WinArch)$($WinConfig)\fnmp\" $DstPath\bin
+copy -Recurse "artifacts\bin\$($WinArch)$($WinConfig)\fnmpapi.dll" $DstPath\bin
 
 New-Item -Path $DstPath\symbols -ItemType Directory > $null
-copy "artifacts\bin\$($WinArch)$($WinConfig)\fnmp.pdb"   $DstPath\symbols
+copy "artifacts\bin\$($WinArch)$($WinConfig)\fnmp.pdb" $DstPath\symbols
+copy "artifacts\bin\$($WinArch)$($WinConfig)\fnmpapi.pdb" $DstPath\symbols
 
 New-Item -Path $DstPath\tools -ItemType Directory > $null
 copy ".\tools\common.ps1" $DstPath\tools

